lib/json-schema/util/uri.rb

```diff
@@ -11,6 +16,11 @@ def self.normalized_uri(uri)
         end
         uri
       end
+
+      def self.normalized_uri(uri)
+        instance.cache ||= {}
+        instance.cache[uri] ||= instance.normalized_uri(uri)
```

Comment:
The caching here is totally unsafe in a multithreaded environment. You're going to end up with race conditions. But the side effect of that is (I admit) fairly minor here because the caller will always get the correct result.

However, I've had a long think about whether caching is the right approach (since you raised this pull request) and I think you're probably right to add it.